File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/IntradayBarElementDateTime.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/IntradayBarElementDateTime.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.Schema;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.Datetime;
import java.text.SimpleDateFormat;
import java.lang.StringBuilder;

public class IntradayBarElementDateTime extends Element
{
    private final Datetime _date;
    
    IntradayBarElementDateTime(Datetime date)
    {
    	this._date = date;
    }
    
    public Name name() throws Exception
    {
    	return new Name("time");
    }
    
    public int numValues()
    {
    	return 1;
    }
    
    public int numElements()
    {
    	return 0;
    }
    
    public boolean isComplexType()
    {
    	return false;
    }
    
    public boolean isArray()
    {
    	return false;
    }
    
    public boolean isNull()
    {
    	return false;
    }

    public Schema.Datatype datatype()
    {
    	return Schema.Datatype.DATETIME;
    }
    
    public boolean hasElement(String name)
    {
    	return false;
    }
    
    public boolean hasElement(String name, boolean excludeNullElements)
    {
    	return false;
    }
    
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
    	return super.prettyPrintHelper(tabIndent, com.bloomberglp.blpapi.DisplayFormats.Intraday_FormatDatetime(this._date));
    }
    
    public Object getValue()
    {
    	return this._date;
    }
    
    public Object getValue(int index) throws Exception
    {
    	if(index == 0)
    		return this.getValue();
    	else
    		return super.getValue();
    }
    
    public Datetime getValueAsDate()
    {
        return this._date;
    }

    public Datetime getValueAsDate(int index) throws Exception
    {
        if (index == 0)
            return this.getValueAsDate();
        else
            return super.getValueAsDate(index);
    }

    public Datetime getValueAsDatetime(int index) throws Exception
    {
        if (index == 0)
            return this.getValueAsDatetime();
        else
            return super.getValueAsDatetime(index);
    }

    public Datetime getValueAsDatetime()
    {
        return this._date;
    }

    public Datetime getValueAsTime()
    {
        return this._date;
    }

    public Datetime getValueAsTime(int i) throws Exception
    {
        if (i == 0)
            return this.getValueAsTime();
        else
            return super.getValueAsTime(i);
    }

    public String getValueAsString()
    {
    	SimpleDateFormat df = new SimpleDateFormat();
    	df.applyPattern("yyyy-MM-dd'T'HH:mm:ss.SSS");
    	String strDate = df.format(this._date.calendar().getTime());
    	return strDate;
    }

    public String getValueAsString(int i) throws Exception
    {
        if (i == 0)
            return this.getValueAsString();
        else
            return super.getValueAsString(i);
    }
}

File: BEmu_cpp/src/IntradayBarRequest/IntradayBarElementDateTime.cpp
//------------------------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/IntradayBarRequest/IntradayBarElementDateTime.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------------------------

#include "IntradayBarRequest/IntradayBarElementDateTime.h"
#include "BloombergTypes/Name.h"
#include "BloombergTypes/ElementPtr.h"
#include "Types/DisplayFormats.h"

namespace BEmu
{
	namespace IntradayBarRequest
	{
		IntradayBarElementDateTime::IntradayBarElementDateTime(const std::string& name, const Datetime& value) : CanConvertToStringType(value)
		{
			this->_name = name;
			this->_value = value;
		}

		IntradayBarElementDateTime::~IntradayBarElementDateTime()
		{
		}

		Name IntradayBarElementDateTime::name() const
		{
			return Name(this->_name.c_str());
		}

		size_t IntradayBarElementDateTime::numValues() const
		{
			return 1;
		}

		size_t IntradayBarElementDateTime::numElements() const
		{
			return 0;
		}

		bool IntradayBarElementDateTime::isArray() const
		{
			return false;
		}

		bool IntradayBarElementDateTime::isComplexType() const
		{
			return false;
		}

		Datetime IntradayBarElementDateTime::getValueAsDatetime(int index) const
		{
			if(index == 0)
				return this->_value;
			else
				throw elementPtrEx;
		}

		const char * IntradayBarElementDateTime::getValueAsString(int index) const
		{
			if(index == 0)
				return this->ValueAsString().c_str();
			else
				throw elementPtrEx;
		}

		std::ostream& IntradayBarElementDateTime::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string strvalue(DisplayFormats::IntradayRequests_FormatDatetime(this->_value));
			this->prettyPrintHelper(stream, level, spacesPerLevel, strvalue);
			return stream;
		}
	}
}